File: platform/pylib/scout_common/observability/metrics/histogram.py
# platform/pylib/scout_common/observability/metrics/histogram.py
"""Histogram metrics for tracking distributions."""

import math
import threading
from typing import Dict, List, Optional


# Default histogram buckets (in seconds)
# Suitable for HTTP request latencies: 5ms, 10ms, 25ms, 50ms, 100ms, 250ms, 500ms, 1s, 2.5s, 5s, 10s
DEFAULT_BUCKETS = [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]


class Histogram:
    """Histogram tracks the distribution of values.
    
    Used for tracking latencies, request sizes, etc.
    
    Example:
        histogram = Histogram("http_request_duration_seconds", "HTTP request latency")
        histogram.observe(0.042)  # 42ms
        histogram.observe(0.156)  # 156ms
        print(f"Count: {histogram.count}, Sum: {histogram.sum}")
        print(f"Buckets: {histogram.buckets}")
    """

    def __init__(
        self,
        name: str,
        help: str,
        labels: Optional[Dict[str, str]] = None,
        buckets: Optional[List[float]] = None
    ):
        """Initialize histogram.
        
        Args:
            name: Metric name
            help: Help description
            labels: Label key-value pairs
            buckets: Histogram buckets (defaults to DEFAULT_BUCKETS)
        """
        self._name = name
        self._help = help
        self._labels = labels or {}
        
        # Sort buckets and add +Inf
        if buckets is None:
            buckets = DEFAULT_BUCKETS.copy()
        self._buckets = sorted(buckets) + [math.inf]
        self._counts = [0] * len(self._buckets)
        
        self._sum = 0.0
        self._count = 0
        self._lock = threading.RLock()

    def observe(self, value: float) -> None:
        """Record a value.
        
        Args:
            value: Value to observe
        """
        with self._lock:
            self._sum += value
            self._count += 1
            
            # Find bucket and increment
            for i, bucket in enumerate(self._buckets):
                if value <= bucket:
                    self._counts[i] += 1

    @property
    def count(self) -> int:
        """Get total number of observations."""
        with self._lock:
            return self._count

    @property
    def sum(self) -> float:
        """Get sum of all observed values."""
        with self._lock:
            return self._sum

    @property
    def buckets(self) -> Dict[float, int]:
        """Get histogram buckets with counts."""
        with self._lock:
            return dict(zip(self._buckets, self._counts))

    @property
    def name(self) -> str:
        """Get histogram name."""
        return self._name

    @property
    def help(self) -> str:
        """Get histogram help text."""
        return self._help

    @property
    def labels(self) -> Dict[str, str]:
        """Get histogram labels."""
        return self._labels.copy()

    def __repr__(self) -> str:
        """Get debug representation."""
        labels_str = ", ".join(f"{k}={v}" for k, v in self._labels.items())
        return f"Histogram({self._name}{{{labels_str}}}, count={self._count}, sum={self._sum})"

File: platform/pkg/observability/metrics/histogram.go
// platform/pkg/observability/metrics/histogram.go
package metrics

import (
	"math"
	"sync"
)

// Histogram tracks the distribution of values.
// Used for tracking latencies, request sizes, etc.
type Histogram interface {
	// Observe records a value
	Observe(value float64)

	// Count returns the total number of observations
	Count() uint64

	// Sum returns the sum of all observed values
	Sum() float64

	// Buckets returns the histogram buckets with counts
	Buckets() map[float64]uint64

	// Name returns the histogram name
	Name() string

	// Labels returns the histogram labels
	Labels() map[string]string
}

// histogram implements Histogram.
type histogram struct {
	name    string
	help    string
	labels  map[string]string
	buckets []float64
	counts  []uint64
	sum     float64
	count   uint64
	mu      sync.RWMutex
}

// DefaultBuckets are the default histogram buckets (in seconds).
// Suitable for HTTP request latencies: 5ms, 10ms, 25ms, 50ms, 100ms, 250ms, 500ms, 1s, 2.5s, 5s, 10s
var DefaultBuckets = []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10}

// NewHistogram creates a new histogram with default buckets.
func NewHistogram(name, help string, labels map[string]string) Histogram {
	return NewHistogramWithBuckets(name, help, labels, DefaultBuckets)
}

// NewHistogramWithBuckets creates a new histogram with custom buckets.
func NewHistogramWithBuckets(name, help string, labels map[string]string, buckets []float64) Histogram {
	if labels == nil {
		labels = make(map[string]string)
	}

	// Sort buckets and add +Inf
	sortedBuckets := make([]float64, len(buckets)+1)
	copy(sortedBuckets, buckets)
	sortedBuckets[len(buckets)] = math.Inf(1)

	return &histogram{
		name:    name,
		help:    help,
		labels:  labels,
		buckets: sortedBuckets,
		counts:  make([]uint64, len(sortedBuckets)),
		sum:     0,
		count:   0,
	}
}

// Observe records a value.
func (h *histogram) Observe(value float64) {
	h.mu.Lock()
	defer h.mu.Unlock()

	h.sum += value
	h.count++

	// Find bucket and increment
	for i, bucket := range h.buckets {
		if value <= bucket {
			h.counts[i]++
		}
	}
}

// Count returns the total number of observations.
func (h *histogram) Count() uint64 {
	h.mu.RLock()
	defer h.mu.RUnlock()
	return h.count
}

// Sum returns the sum of all observed values.
func (h *histogram) Sum() float64 {
	h.mu.RLock()
	defer h.mu.RUnlock()
	return h.sum
}

// Buckets returns the histogram buckets with counts.
func (h *histogram) Buckets() map[float64]uint64 {
	h.mu.RLock()
	defer h.mu.RUnlock()

	result := make(map[float64]uint64)
	for i, bucket := range h.buckets {
		result[bucket] = h.counts[i]
	}
	return result
}

// Name returns the histogram name.
func (h *histogram) Name() string {
	return h.name
}

// Labels returns the histogram labels.
func (h *histogram) Labels() map[string]string {
	return h.labels
}